### DISCUSSION TITLE
Is it possible to predownload ports in emsdk docker?

### DISCUSSION THREAD

**Discussion by ericoporto** [emscripten-core/emscripten#16165] [Q&A] [answered]
I am trying to make an emscripten dockerfile, I am going to cache this docker image and run the builds on this docker image.

I want to also pre-download the Emscripten ports. 

Is there any command line command that can do this? My docker image uses emsdk from here (`FROM emscripten/emsdk:3.1.2`): https://hub.docker.com/r/emscripten/emsdk

**Comment by kleisauke** [comment, marked as answer]
I think you can extend the existing Docker image and run `embuilder.py build USER` to download (and build) all ports. For example:
```dockerfile
FROM emscripten/emsdk:3.1.3

RUN embuilder.py build USER
```
```bash
$ docker build -t emscripten-ports . 
```

(note that this currently fails for zlib due to https://github.com/emscripten-core/emscripten/issues/16164)

**Reply by ericoporto** [reply]
That (the Unexpected hash) is actually what prompted me to go this way, at least for me - I am on 3.1.2, so it was a surprise too. I don't really need zlib, if it was possible to select the ports...

Sorry, what's the USER keyword there? Is it possible to specify which ports should build? I can't tell from reading the embuilder.py source code...

**Reply by kleisauke** [reply]
Issuing `embuilder.py build USER` causes every port to be built. `embuilder.py` accepts multiple arguments, so you can just do `embuilder.py build freetype harfbuzz` (for example).

**Reply by ericoporto** [reply]
thanks, did just that and it worked! 😄

**Reply by ericoporto** [reply]
(hey, is it possible to pre-generate libc so I can skip `generating system library: sysroot/lib/wasm32-emscripten/thinlto/libc.a.` ?)

**Reply by ericoporto** [reply]
The specific problem being the flag to force `thinlto`

from embuilder:

```
  if args.lto:
    settings.LTO = "full"
```

I don't need full...

**Reply by kleisauke** [reply]
For regular/full LTO, you could use `embuilder.py build libc --lto`. For thin LTO (which is not heavily tested), there isn't such argument. Perhaps we should extend the `--lto` setting of `embuilder.py` by accepting value arguments (e.g., `--lto=thin`), just like `emcc`.
https://github.com/emscripten-core/emscripten/blob/e415e01e8e6ddc7c453e97e557ea0e2cf37d5d23/emcc.py#L2996-L3000
https://github.com/emscripten-core/emscripten/blob/e415e01e8e6ddc7c453e97e557ea0e2cf37d5d23/embuilder.py#L153-L154

(Or allowing something like `EMCC_CFLAGS="-flto=thin" embuilder.py build libc`)
